devel/shtk/files/shtk.1
.\" Copyright 2012 Google Inc.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions are
.\" met:
.\"
.\" * Redistributions of source code must retain the above copyright
.\"   notice, this list of conditions and the following disclaimer.
.\" * Redistributions in binary form must reproduce the above copyright
.\"   notice, this list of conditions and the following disclaimer in the
.\"   documentation and/or other materials provided with the distribution.
.\" * Neither the name of Google Inc. nor the names of its contributors
.\"   may be used to endorse or promote products derived from this software
.\"   without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
.\" DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
.\" THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
.\" OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.Dd August 15, 2012
.Dt SHTK 1
.Os
.Sh NAME
.Nm shtk
.Nd interface to the Shell Toolkit
.Sh SYNOPSIS
.Nm
build
.Op Fl m Ar main_function
.Op Fl o Ar output_file
.Op Fl s Ar shell
.Op Ar input_file
.Nm
version
.Sh DESCRIPTION
The Shell Toolkit, or
.Nm
for short, is a collection of modules written in
.Xr sh 1
that provide common functionality to simplify the implementation of complex
shell scripts.
.Pp
The tool described in this page, also named
.Nm ,
is a program that accompanies the Shell Toolkit and provides functionality that
may be necessary to use the toolkit.
.Pp
The interface of
.Nm
is that of a program with commands as shown in the synopsis section.
The following subsections describe the various commands supported by the tool.
.Ss The build command
The build command takes a script that uses
.Nm
modules and generates an executable script with the necessary boilerplate code
to be able to locate said modules.
.Pp
The
.Ar input_file
argument specifies the path to the file to process.
This file should end with an
.Sq .sh
extension, in which case the build command will proceed to generate a binary of
the same name without any extension.
If the extension is not present, then the name of the output file must be
provided by specifying the
.Fl o
flag.
The input file can also be
.Sq - ,
in which case the code is read from the standard input.
.Pp
The following options are supported:
.Bl -tag -width XXXX
.It Fl m Ar main_function
Name of the function that implements the entry point of the program.
.Pp
All scripts that use
.Nm
should implement a main function and should avoid placing any code at the top
level of the file.
The main function is automatically executed by the generated script.
If the value of
.Ar main_function
is empty, then no call to main is issued which may be necessary if the script
already has a call to main for whatever reason (e.g. the script also uses
.Xr atf-sh 1 ) .
.Pp
Default:
.Sq main .
.It Fl o Ar output_file
Path to the output file to generate.
.Pp
Default: if
.Ar input_file
ends in
.Sq .sh ,
the output file has the same name as the input file but without the extension.
Otherwise this flag must be specified.
.It Fl s Ar shell
Path to the shell interpreter to write in the shebang line of the file.
.Pp
Default:
.Pa @SHTK_SHELL@ .
.El
.Ss The version command
The version command prints the name of the package and its version to the
standard output.
This command always exits successfully.
.Sh ENVIRONMENT
.Bl -tag -width XXXX
.It Va SHTK_MODULESDIR
Overrides the built-in location of the modules.
.It Va SHTK_SHELL
Overrides the built-in location of the default shell interpreter.
.El
.Sh FILES
.Bl -tag -width XXXX
.It Pa @SHTK_MODULESDIR@/
Location of the
.Nm
modules.
.El
.Sh SEE ALSO
Unfortunately, there currently is no standalone documentation of the
.Nm .
The documentation exists, but is in the form of docstrings next to the
functions they describe.
Such docstrings document the purpose of the function, the parameters it
takes, its return value (if any) and any side-effects.
.Pp
Please look at the files in
.Pa @SHTK_MODULESDIR@
for additional details.
.Sh AUTHORS
The
.Nm
package was developed by
.An Julio Merino
.Aq jmmv@NetBSD.org .


doc/NEWS
# $NetBSD: NEWS,v 1.4 2008/02/27 16:49:34 rillig Exp $

PKGSRC NEWS
===========

  This file contains information about important infrastructural changes in
  pkgsrc. The intended readers are the pkgsrc developers as well as anyone
  tracking pkgsrc-current.

SKIP_PORTABILITY_CHECK is obsolete [rillig 20080220]
	After being deprecated for about 16 months, this early misdesign
	has finally been obsoleted and removed from pkgsrc.

Check for stripped/unstripped binaries [rillig 20080213]
	A new check has been added in mk/check/check-stripped.mk to
	assure that the installed binaries conform to the setting of the
	INSTALL_UNSTRIPPED variable. It is not enabled by default due to
	lack of testing.

OpenSSL update [tnn 20080118]
	OpenSSL in pkgsrc switched from the 0.9.7 to the 0.9.8 branch. To
	accomodate for this, 596 packages dependent on openssl had their
	revisions bumped.

pkgsrc-2007Q4 released [agc 20080105]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2008/01/05/0000.html

Simplifying meta-packages [rillig 20071219]
	The variable META_PACKAGE was introduced. This simplifies creation
	and maintenance of packages that do not install files.

Replace a package with different name [gdt 20071107]
	The variable OLDNAME was introduced for "make replace" to
	replace a package with a name other than the new one, via
	"make replace OLDNAME=oldpkgname".  Read the announcement at:
	http://mail-index.netbsd.org/tech-pkg/2007/11/07/0002.html
	
pkgsrc-2007Q3 released [agc 20071015]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/10/15/0000.html

Features framework [jlam 20070907]
	A "system features" framework has been added. It will be used to
	automatically supply missing "basic" headers and libraries for old
	systems. This is supported through the USE_FEATURES variable.

X.org 6.9 removed [joerg 20070817]
	monolithic X.org 6.9 was removed from pkgsrc. It was obsoleted by the
	modular X.org (7.x) packages. Transitioning users should remove all old
	X.org packages and set X11_TYPE=modular in mk.conf.

Emulator framework [jlam 20070729]
	A new emulator framework has been added. This will handle all
	binary-only packages that require emulation on the native OS.
	Please see pkgsrc/mk/emulator/README.

audit-packages merged with pkg_install [adrianp 20070714]
	The utility for auditing installed packages for security problems,
	audit-packages, has been rewritten in C. It is now part of the
	pkgtools/pkg_install package.
	
pkgsrc-2007Q2 released [agc 20070701]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/07/01/0000.html

New framework for bulk builds [joerg 20070619]
	pbulk, the new automated pkgsrc bulk build framework was imported.

pkgsrc-2007Q1 released [agc 20070419]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/04/19/0000.html

pkgsrc-2006Q4 released [agc 20070104]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2007/01/04/0000.html

modular X.org imported [joerg 20061114]
	The modular X.org packages were imported.

pkgsrc-2006Q3 released [agc 20061001]
	A new stable branch was released. Read the announcement at:
	http://mail-index.netbsd.org/netbsd-announce/2006/10/01/0000.html


devel/shtk/files/shtk.sh
#! @SHTK_SHELL@
# Copyright 2012 Google Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
# * Redistributions of source code must retain the above copyright
#   notice, this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright
#   notice, this list of conditions and the following disclaimer in the
#   documentation and/or other materials provided with the distribution.
# * Neither the name of Google Inc. nor the names of its contributors
#   may be used to endorse or promote products derived from this software
#   without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

# \file shtk.sh
# Entry point and main program logic.
#
# For simplicity reasons, we cannot rely on any of our own modules to
# implement this file.  Doing so, while possible, would complicate the
# creation of the final shtk script for no real advantage.

set -e


# Location of the shtk modules.
: ${SHTK_MODULESDIR:="@SHTK_MODULESDIR@"}


# Default shell to use when generating scripts.
: ${SHTK_SHELL:="@SHTK_SHELL@"}


# Version of the package.
SHTK_VERSION="@SHTK_VERSION@"


# Base name of the running script.
_ProgName="${0##*/}"


# Prints a runtime error and exits.
#
# \param ... The message to print.  Can be provided as multiple words and, in
#     that case, they are joined together by a single whitespace.
error() {
    echo "${_ProgName}: E: $*" 1>&2
    exit 1
}


# Prints an usage error and exits.
#
# \param ... The message to print.  Can be provided as multiple words and, in
#     that case, they are joined together by a single whitespace.
usage_error() {
    echo "${_ProgName}: E: $*" 1>&2
    echo "Type 'man ${_ProgName}' for help" 1>&2
    exit 1
}


# Command to build a script that uses shtk libraries.
#
# \params ... Options and arguments to the command.
shtk_build() {
    local main=main
    local output=
    local shell="${SHTK_SHELL}"

    while getopts ':m:o:s:' arg "${@}"; do
        case "${arg}" in
            m)  # Main function name.
                main="${OPTARG}"
                ;;

            o)  # Output file.
                output="${OPTARG}"
                ;;

            s)  # Shell to use.
                shell="${OPTARG}"
                ;;

            \?)
                usage_error "Unknown option -${OPTARG}"
                ;;
        esac
    done
    shift $((${OPTIND} - 1))

    [ ${#} -eq 1 ] || usage_error "build takes one argument only"

    local input="${1}"; shift
    case "${input}" in
        *.sh)
            [ -n "${output}" ] || output="$(echo ${input} | sed -e 's,\.sh$,,')"
            ;;

        *)
            [ -n "${output}" ] || usage_error "Input file should" \
                "end in .sh or you must specify -o"
            ;;
    esac

    [ "${input}" = - -o -e "${input}" ] || error "Cannot open ${input}"

    # Note that we use the built-in value of SHTK_MODULESDIR unconditionally
    # instead of what the environment says to avoid possible side-effects that
    # would be easy to debug.
    sed -e "s,__SHTK_MODULESDIR__,@SHTK_MODULESDIR@,g" \
        -e "s,__SHTK_SH__,${shell},g" \
        "${SHTK_MODULESDIR}/bootstrap.subr" \
        | grep -v '^#[^!].*' | grep -v '^#$' >"${output}.tmp"
    if [ "${input}" = - ]; then
        cat >>"${output}.tmp"
    else
        cat "${input}" >>"${output}.tmp"
    fi
    [ -z "${main}" ] || echo "${main} \"\${@}\"" >>"${output}.tmp"
    chmod +x "${output}.tmp"
    if ! mv "${output}.tmp" "${output}"; then
        rm -f "${output}.tmp"
        error "Failed to create ${output}"
    fi
}


# Gets version information about shtk.
shtk_version() {
    [ ${#} -eq 0 ] || usage_error "version does not take any arguments"

    echo "shtk ${SHTK_VERSION}"
}


# Entry point to the program.
#
# \param ... Command-line arguments to be processed.
#
# \return An exit code to be returned to the user.
shtk_main() {
    [ ${#} -ge 1 ] || usage_error "No command specified"

    local exit_code=0

    local command="${1}"; shift
    case "${command}" in
        build|version)
            "shtk_${command}" "${@}" || exit_code="${?}"
            ;;

        *)
            usage_error "Unknown command ${command}"
            ;;
    esac

    return "${exit_code}"
}


shtk_main "${@}"
